[framework] Adds publisher module

damirka opened the pull request:
Finalizes the experiment myself and @amnn started; and proposes a `publisher` module.

The goal of this module is to enable recognition of the Publisher authority. Potentially it could be used for:
1. Setting rules / registering a type in a system, such as metadata, AFTER the module was published 
2. Authorizing admin actions (effectively replacing AdminCap which is often added manually)

sblackshear commented on crates/sui-framework/sources/publisher.move:
```diff
@@ -0,0 +1,145 @@
+            let left = vector::borrow<u8>(this, i);
+            let right = vector::borrow<u8>(their, i);
+
+            if (left == &ASCII_COLON && right == &ASCII_COLON) {
```
I would be inclined to put this logic for module name parsing in typename, since this surely won't be the last usage.

damirka replied:
Agreed! Makes a lot of sense!

damirka replied:
On the other hand I wouldn't make it a blocker for now since it will require changes in the std in the move repo.

amnn replied:
Should we mirror `stdlib` modules under the `sui` package, so we have a home for things that belong in a place like stdlib before we upstream them?

damirka replied:
@amnn @sblackshear 
One question I don't have an answer to is how can we know the length of the address type in Move? This module currently relies on addresses being 20 bytes (40 ASCII symbols). Should we include logic for 32-byte addresses as well?

amnn replied:
It's agood question, I think we should consider introducing a native `std::address::length()` to expose the `AccountAddress::LENGTH` constant within the VM (and this can supersede our own `sui::address::length()`.

damirka replied:
Added `std::address::length` with a hardcoded const.

amnn commented on crates/sui-framework/sources/publisher.move:
```diff
@@ -0,0 +1,145 @@
+    }
+
+    /// Claim a Publisher object.
+    /// Requires a One-Time-Witness to
```
nit: unfinished sentence?

amnn commented on crates/sui-framework/sources/publisher.move:
```diff
@@ -0,0 +1,145 @@
+    /// Check whether type belongs to the same package as the publisher object.
+    public fun is_package<T>(publisher: &Publisher): bool {
+        let this = ascii::as_bytes(&publisher.type);
+        let their = ascii::as_bytes(type_name::borrow_string(&type_name::get<T>()));
```
nit (extremely optional): I like `this` and `that` for grammatical symmetry, but this literally does not matter at all.

damirka replied:
Agreed, will fix in a bit.

amnn commented on crates/sui-framework/sources/publisher.move:
```diff
@@ -0,0 +1,145 @@
+    use sui::test_scenario::{Self as test, Scenario, ctx};
+
+    /// OTW for the test_publisher module
+    struct TEST_OTW has drop {}
```
How does his end up being interpreted as a OTW, given that its name doesn't match, is it some test-mode magic?  It would be good to document what's going on here, in case someone else needs to do the same thing.

damirka replied:
Haha. It's a workaround because I wasn't able to create an OTW in a `#[test]` function. If you look at how it's being used, there's a `test_claim` function instead of `claim` which allows any struct really as long as it has drop.

amnn reviewed (commented):
I like it!  Just waiting to finish the conversation about a home for things that should eventually belong in move stdlib before final stamp.

damirka commented:
@amnn @sblackshear on the topic: do you think we should rename it into `PublisherCap` instead of `Publisher`?

amnn commented:
> do you think we should rename it into PublisherCap instead of Publisher?

I think technically it's a witness (i.e. it witnessed the package being published), rather than a capability.

damirka commented:
@amnn @sblackshear 

I've addressed your comments above. Few changes:

- adds `std::address` library with a `length()` function
- adds address getter + module name getter in `std::type_name`
- does a little cleanup in our `sui::bcs` library to switch to `sui::address` instead of using const for ADDRESS length

amnn commented on crates/sui-framework/deps/move-stdlib/sources/type_name.move:
```diff
@@ -24,6 +29,47 @@ module std::type_name {
+        let str_bytes = ascii::as_bytes(&self.name);
+        let addr_bytes = vector[];
+
+        // Reverse-read string from the last byte pushing the value
```
I don't understand why we are reverse reading the address here?  IIUC, this means `addr_bytes` will be the reverse of the address prefix of `str_bytes`.

damirka replied:
yrros esrever ni gnikniht saw daeh ym

amnn commented on crates/sui-framework/deps/move-stdlib/sources/type_name.move:
```diff
@@ -24,6 +29,47 @@ module std::type_name {
+            }
+        };
+
+        vector::reverse(&mut module_name);
```
Maybe it's just late, and I'm being slow -- but I think this is also going to produce the reverse of the ordering that I expect.  If you had some tests for these new functions, it would help settle my mind that I am indeed being silly.

damirka replied:
Sorry, it's me being dumb. Ofc if we read byte per byte and push back it should work as expected.

amnn commented on crates/sui-framework/sources/publisher.move:
```diff
@@ -7,34 +7,37 @@
     struct Publisher has key, store {
         id: UID,
-        type: String
+        type: TypeName,
```
Is there a reason to keep the `type` here?  I only see `package` and `module_name` being used.

damirka replied:
Good question. It's good for visibility in a sense that people know the full OTW. But on the other hand we could remove it as unused.

amnn replied:
We can probably skip it, right? The OTW can only be one thing, which you can always derive from the moduel name and address.

damirka replied:
Addressed your comment (and other ones).

amnn reviewed (commented):
The Publisher changes look good to me, but unless I'm missing something (very possible), I think `type_name::get_address` and `type_name::get_module` produce reversed outputs compared to what I would expect (and could benefit from some tests) -- could you take a look at those?

amnn commented on crates/sui-framework/sources/publisher.move:
```diff
@@ -103,6 +111,10 @@ module sui::test_publisher {
 
+        assert!(&ascii::string(b"0000000000000000000000000000000000000002") == publisher::package(&pub), 0);
+
+        std::debug::print(publisher::package(&pub));
```
Is the `print` intentional here?

damirka replied:
Nope. Just removed.

amnn commented on crates/sui-framework/sources/publisher.move:
```diff
@@ -103,6 +111,10 @@ module sui::test_publisher {
         assert!(publisher::is_package<CustomType>(&pub), 0);
         assert!(publisher::is_package<Scenario>(&pub), 0);
 
+        assert!(&ascii::string(b"0000000000000000000000000000000000000002") == publisher::package(&pub), 0);
```
nit: These tests probably belong in the stdlib rather than here, as that's where most of the logic is -- but very happy to see them!

damirka replied:
That's a tricky part. We don't have tests in the stdlib and I never dared to add one. Guess it's a bigger question on whether it's okay to have tests for stdlib or there's a better place for them where they live?

amnn replied:
Yeah, interesting -- maybe @sblackshear / @tnowacki has context on why there seems to not be `#[test]`s in the stdlib?

amnn replied:
(Any case, not a blocker for this PR)

amnn reviewed (approved):
Thanks @damirka !
